plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Outline.hs

```diff
@@ -0,0 +1,103 @@
+                                                    useWithStaleFast)
+import           Development.IDE.Types.Location    (toNormalizedFilePath')
+import           Distribution.Fields.Field         (Field (Field, Section),
+                                                    Name (Name),
+                                                    SectionArg (SecArgName, SecArgOther, SecArgStr))
+import           Distribution.Parsec.Position      (Position)
+import           Ide.Plugin.Cabal.Completion.Types (ParseCabalFields (..),
+                                                    cabalPositionToLSPPosition)
+import           Ide.Plugin.Cabal.Orphans          ()
+import           Ide.Types                         (PluginMethodHandler)
+import           Language.LSP.Protocol.Message     (Method (..))
+import           Language.LSP.Protocol.Types       (DocumentSymbol (..))
+import qualified Language.LSP.Protocol.Types       as LSP
+
+
+moduleOutline :: PluginMethodHandler IdeState Method_TextDocumentDocumentSymbol
+moduleOutline ideState _ LSP.DocumentSymbolParams {_textDocument = LSP.TextDocumentIdentifier uri} =
+  case LSP.uriToFilePath uri of
+    Just (toNormalizedFilePath' -> fp) -> do
+      mFields <- liftIO $ runIdeAction "cabal-plugin.fields" (shakeExtras ideState) (useWithStaleFast ParseCabalFields fp)
+      case fmap fst mFields of
+        Just fieldPositions -> pure $ LSP.InR (LSP.InL allSymbols)
+          where
+            allSymbols = mapMaybe documentSymbolForField fieldPositions
+        Nothing -> pure $ LSP.InL []
+    Nothing -> pure $ LSP.InL []
+
+-- | Creates a DocumentSumbol object for the
+--   cabal AST, without displaying fieldLines and
+--   displaying Section name and SectionArgs in one line
+documentSymbolForField :: Field Position -> Maybe DocumentSymbol
+documentSymbolForField (Field (Name pos fieldName) _) =
+  Just
+    (defDocumentSymbol range)
+      { _name = decodeUtf8 fieldName,
+        _kind = LSP.SymbolKind_Field,
+        _children = Nothing
+      }
+  where
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` decodeUtf8 fieldName
+documentSymbolForField (Section (Name pos fieldName) sectionArgs fields) =
+  Just
+    (defDocumentSymbol range)
+      { _name = joinedName,
+        _kind = LSP.SymbolKind_Object,
+        _children =
+          Just
+            (mapMaybe documentSymbolForField fields)
+      }
+  where
+    joinedName = decodeUtf8 fieldName <> " " <> joinedNameForSectionArgs sectionArgs
+    range = cabalPositionToLSPRange pos `addNameLengthToLSPRange` joinedName
+
+joinedNameForSectionArgs :: [SectionArg Position] -> T.Text
+joinedNameForSectionArgs sectionArgs = joinedName
+  where
+    joinedName = T.unwords $ map getName sectionArgs
+
+    getName :: SectionArg Position -> T.Text
+    getName (SecArgName _ identifier)  = decodeUtf8 identifier
+    getName (SecArgStr _ quotedString) = decodeUtf8 quotedString
+    getName (SecArgOther _ string)     = decodeUtf8 string
```

Comment:
Docs please here as well, what does this do? Bonus points for practical examples! A better name perhaps would be `prettySectionArgs`

This function should likely be a part of https://github.com/haskell/haskell-language-server/blob/master/plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completion/CabalFields.hs#L12